// @progress/kendo-react-dropdowns/dist/npm/common/VirtualScroll.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var maxHeightIE = 1533915;
var addPageChange = function (state, skip, take) {
    state.events.push({
        type: 'onPageChange',
        page: {
            skip: skip,
            take: take
        }
    });
};
/**
 * @hidden
 */
var VirtualScroll = /** @class */ (function () {
    function VirtualScroll() {
        var _this = this;
        this.containerHeight = 0;
        this.skip = 0;
        this.total = 0;
        this.enabled = false;
        this.pageSize = 0;
        this.itemHeight = 0;
        this.hidden = false;
        this.prevScrollPos = 0;
        this.listTranslate = 0;
        this.scrollSyncing = false;
        this.scrollerRef = function (element) {
            if (element) {
                var vs = _this;
                vs.container = element;
                element.setAttribute('unselectable', 'on');
                setTimeout(vs.calculateScrollSizes.bind(vs), 0);
            }
        };
        this.calculateScrollSizes = function () {
            _this.scrollSyncing = true;
            var heightChanged = false;
            _this.itemHeight = _this.list ? _this.list.children[0].offsetHeight : _this.itemHeight;
            _this.containerHeight = Math.min(maxHeightIE, _this.itemHeight * _this.total);
            var newHeight = _this.containerHeight;
            heightChanged = _this.scrollElement && _this.scrollElement.style.height !== newHeight + 'px';
            if (heightChanged) {
                _this.scrollElement.style.height = newHeight + 'px';
            }
            _this.scrollSyncing = false;
            return heightChanged;
        };
        this.scrollHandler = this.scrollHandler.bind(this);
    }
    VirtualScroll.prototype.initVirtualization = function (virtual, state) {
        this.enabled = virtual !== undefined;
        if (virtual !== undefined) {
            this.skip = virtual.skip;
            this.pageSize = virtual.pageSize;
            this.total = virtual.total;
            if (state !== undefined) {
                state.data.skip = virtual.skip;
            }
        }
    };
    VirtualScroll.prototype.updateVirtualization = function (next, current, state) {
        if (next !== undefined && current !== undefined && (current.total !== next.total || next.skip === 0)) {
            var settings = Object.assign({}, next, { skip: 0 });
            this.initVirtualization(settings, state);
            if (this.calculateScrollSizes()) {
                this.reset();
            }
        }
        else {
            this.initVirtualization(next, state);
        }
    };
    VirtualScroll.prototype.handleVirtualItemSelect = function (newState, state, scrollElement, index) {
        var skip = state.skip || 0;
        var prevIndex = state.selectedIndex || 0;
        var opened = state.opened && scrollElement;
        var vs = this;
        var take = vs.pageSize;
        var listItemIndex = index - skip;
        if (opened && index === 0) {
            addPageChange(newState, 0, take);
            vs.reset();
        }
        else if (opened && index === vs.total - 1) {
            addPageChange(newState, vs.total - vs.pageSize, take);
            vs.scrollToEnd();
        }
        else if (opened) {
            var pageChange = vs.PageChange;
            vs.PageChange = function (page) {
                addPageChange(newState, page.skip, page.take);
            };
            var scrollHandler = index > prevIndex ? vs.localScrollDown : vs.localScrollUp;
            scrollHandler.call(vs, state.syntheticEvent);
            vs.PageChange = pageChange;
        }
        else if (!opened && index === 0) {
            addPageChange(newState, 0, take);
        }
        else if (!opened && index === vs.total - 1) {
            addPageChange(newState, vs.total - vs.pageSize, take);
        }
        else if (!opened) {
            var newSkip = Math.max(0, skip + (listItemIndex < 0 ? -1 : 1));
            addPageChange(newState, newSkip, take);
        }
        if (!state.opened) {
            vs.reset();
        }
    };
    VirtualScroll.prototype.updateListScroll = function (state, scrollElement, callback) {
        var scrollTop = scrollElement.scrollTop;
        var vs = this;
        var translate = vs.translate;
        var hidden = vs.hidden;
        var skip = state.skip || 0;
        if (scrollTop !== translate || (translate === 0 && scrollTop === 0 && skip > 0)) {
            var resetScroll = function () {
                vs.enabled = false;
                var listOffsetTop = translate > 0 ?
                    translate : (vs.itemHeight * skip);
                scrollElement.scrollTop = listOffsetTop;
                vs.translateTo(translate || listOffsetTop);
                vs.enabled = true;
                callback();
            };
            if (vs.itemHeight !== 0 && !hidden) {
                resetScroll();
            }
            else {
                setTimeout(resetScroll, 0);
            }
        }
    };
    Object.defineProperty(VirtualScroll.prototype, "translate", {
        get: function () {
            return this.listTranslate;
        },
        enumerable: true,
        configurable: true
    });
    VirtualScroll.prototype.changePage = function (skip, e) {
        var newSkip = Math.min(Math.max(0, skip), this.total - this.pageSize);
        if (newSkip !== this.skip) {
            this.PageChange({ skip: newSkip, take: this.pageSize }, e);
        }
    };
    VirtualScroll.prototype.translateTo = function (dY) {
        this.listTranslate = dY;
        if (this.list) {
            this.list.style.transform = 'translateY(' + dY + 'px)';
        }
    };
    VirtualScroll.prototype.reset = function () {
        this.scrollSyncing = true;
        if (this.container) {
            this.container.scrollTop = 0;
        }
        this.translateTo(0);
        this.scrollSyncing = false;
    };
    VirtualScroll.prototype.maxScroll = function () {
        return this.container.scrollHeight - this.container.offsetHeight;
    };
    VirtualScroll.prototype.scrollToEnd = function () {
        if (this.container) {
            var scroll_1 = this.maxScroll();
            this.translateTo(scroll_1);
            this.container.scrollTop = scroll_1;
        }
    };
    VirtualScroll.prototype.localScrollUp = function (e) {
        var height = this.itemHeight;
        var scrollTop = this.container.scrollTop;
        var targetTranslate = this.listTranslate;
        var items;
        var additionalOnTop = scrollTop - targetTranslate;
        if (additionalOnTop > 0) {
            return;
        }
        for (items = 0; items < this.skip; items++) {
            if (targetTranslate + height + additionalOnTop <= scrollTop) {
                break;
            }
            targetTranslate -= height;
        }
        targetTranslate = this.validateTranslate(targetTranslate);
        if (this.skip - items <= 0 && targetTranslate > scrollTop) {
            this.translateTo(0);
            this.changePage(0, e);
            this.container.scrollTop = 0;
            return;
        }
        if (targetTranslate !== this.listTranslate) {
            this.translateTo(targetTranslate);
            this.changePage(this.skip - items, e);
        }
    };
    VirtualScroll.prototype.localScrollDown = function (e) {
        var height = this.itemHeight;
        var scrollTop = this.container.scrollTop;
        var targetTranslate = this.listTranslate;
        var itemsLenght = this.list.children.length;
        var items;
        for (items = 0; items < itemsLenght; items++) {
            if (targetTranslate + height > scrollTop) {
                break;
            }
            targetTranslate += height;
        }
        targetTranslate = this.validateTranslate(targetTranslate);
        if (items >= itemsLenght && this.skip + items >= this.total) {
            this.translateTo(targetTranslate);
            this.changePage(this.total - 1, e);
        }
        else if (targetTranslate !== this.listTranslate) {
            this.translateTo(targetTranslate);
            this.changePage(this.skip + items, e);
        }
    };
    VirtualScroll.prototype.scrollNonStrict = function (e) {
        var floatItemIndex = this.total * this.prevScrollPos / (this.containerHeight);
        var itemIndex = Math.min(Math.floor(floatItemIndex), this.total - 1);
        var targetTranslate = this.containerHeight * floatItemIndex / this.total;
        targetTranslate = this.validateTranslate(targetTranslate);
        this.translateTo(targetTranslate);
        this.changePage(itemIndex, e);
    };
    VirtualScroll.prototype.scrollHandler = function (e) {
        if (!this.enabled || !this.list || this.hidden || this.scrollSyncing) {
            return;
        }
        var scrollTop = this.container.scrollTop;
        var prev = this.prevScrollPos;
        this.prevScrollPos = scrollTop;
        if (scrollTop - prev < 0 && scrollTop > this.listTranslate - this.list.scrollHeight / 10) {
            this.localScrollUp(e);
        }
        else if (scrollTop - prev > 0 && scrollTop < this.listTranslate + this.list.scrollHeight * 2 / 3) {
            this.localScrollDown(e);
        }
        else {
            this.scrollNonStrict(e);
        }
    };
    VirtualScroll.prototype.validateTranslate = function (translate) {
        translate = Math.max(0, translate);
        translate = Math.min(this.containerHeight - this.list.offsetHeight, translate);
        return translate;
    };
    return VirtualScroll;
}());
exports.default = VirtualScroll;
